### PR TITLE
Handling the Custom Nodes and Trusted Location message

### PR DESCRIPTION
**Pull request by jesusalvino**
### Purpose

Handling the Trusted location message when a custom node space is opened.
https://jira.autodesk.com/browse/DYN-5089

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers
@QilongTang 

### FYIs
@RobertGlobant20 @filipeotero

### CODE REVIEW

**Comment by jesusalvino** [comment]
trusted

File: src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs [hunk 1]
```diff
@@ -135,9 +135,10 @@ private void YesButton_Click(object sender, RoutedEventArgs e)
                 if (dynViewModel.PreferenceSettings.AddTrustedLocation(fileTrustWarningViewModel.DynFileDirectoryName))
                     dynViewModel.MainGuideManager.CreateRealTimeInfoWindow(string.Format(Properties.Resources.TrustLocationAddedNotification, fileTrustWarningViewModel.DynFileDirectoryName));
             }
-            if (dynViewModel.CurrentSpaceViewModel.RunSettingsViewModel.Model.RunType != RunType.Manual)
+            // The maximum workspaces always will be Two, The first one for the Dynamo file and the second one for a Custom Node
+            if (dynViewModel.Workspaces[0].RunSettingsViewModel.Model.RunType != RunType.Manual)
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs:139]: I think it's safer to not assume this won't change, and instead check that the workspace is a home workspace.

Comment (QilongTang) [reply src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs:139]: I agree

**Review by QilongTang — APPROVED** [review]

**Comment by QilongTang** [comment]
ALthough this change is good, can you hide the trust warning when switching to a different workspace which is a custom node workspace?

**Comment by jesusalvino** [comment]
> Although this change is good, I don't think we should display trust warning in the bug senario from the first place. Can you hide the trust warning when switching to a different workspace which is a custom node workspace?
trusted2

**Comment by jesusalvino** [comment]
> Although this change is good, I don't think we should display trust warning in the bug senario from the first place. Can you hide the trust warning when switching to a different workspace which is a custom node workspace?

trusted4

File: src/DynamoCoreWpf/Controls/NotificationsControl.xaml.cs [hunk 2]
```diff
@@ -40,9 +40,14 @@ void NotificationsControl_Loaded(object sender, RoutedEventArgs e)
...
-            var hsvm = (HomeWorkspaceViewModel)((DynamoViewModel)DataContext).HomeSpaceViewModel;
+            var hsvm = (HomeWorkspaceViewModel)((DynamoViewModel)DataContext).HomeSpaceViewModel;            
             // Commented this after MAGN - 8423
-            // hsvm.ClearWarning();
+            //hsvm.ClearWarning();
+            if (hsvm.IsHomeSpace && !hsvm.IsCurrentSpace)
+            {
+                hsvm.DynamoViewModel.FileTrustViewModel.ShowWarningPopup = false;
+                hsvm.ClearWarning();
+            }
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Controls/NotificationsControl.xaml.cs:50]: hmm why do we need to touch this piece of code?

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs [hunk 3]
```diff
@@ -270,6 +270,7 @@ internal void UpdateRunStatusMsgBasedOnStates()
             if(IsHomeSpace && !IsCurrentSpace)
             {
                 SetCurrentWarning(NotificationLevel.Mild, string.Empty);
+                DynamoViewModel.FileTrustViewModel.ShowWarningPopup = false;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:273]: From the gif, there will be a flash of trust warning displayed then disappear.. Less desirable, can we hide it even earlier?

File: src/DynamoCoreWpf/Controls/NotificationsControl.xaml.cs (hunk 2 above)
Comment (jesusalvino) [reply src/DynamoCoreWpf/Controls/NotificationsControl.xaml.cs:50]: it forces to hide the untrusted message and bubble when we start to work in a just created custom node workspace, not only when we switch the workspaces.

**Comment by QilongTang** [comment]
@jesusalvino Nice tries but I would love something more granular, for example, you can touch the visibility at https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs#L220 before executing the command to switch tabs. Or you  can subscribe to `ViewingHomespace` property change to determine the visibility of file trust warning. From my point of view, touching visibility after the tab is switched is less desirable to me.

**Comment by jesusalvino** [comment]
> @jesusalvino Nice tries but I would love something more granular, for example, you can touch the visibility at https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs#L220 before executing the command to switch tabs. Or you can subscribe to `ViewingHomespace` property change to determine the visibility of file trust warning. From my point of view, touching visibility after the tab is switched is less desirable to me.

Done

trustedx

File: src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs [hunk 4]
```diff
@@ -134,10 +139,10 @@ private void YesButton_Click(object sender, RoutedEventArgs e)
             {
                 if (dynViewModel.PreferenceSettings.AddTrustedLocation(fileTrustWarningViewModel.DynFileDirectoryName))
                     dynViewModel.MainGuideManager.CreateRealTimeInfoWindow(string.Format(Properties.Resources.TrustLocationAddedNotification, fileTrustWarningViewModel.DynFileDirectoryName));
-            }
-            if (dynViewModel.CurrentSpaceViewModel.RunSettingsViewModel.Model.RunType != RunType.Manual)
+            }            
+            if ((dynViewModel.HomeSpaceViewModel as HomeWorkspaceViewModel).RunSettingsViewModel.Model.RunType != RunType.Manual)
             {
-                dynViewModel.Model.CurrentWorkspace.RequestRun();
+                (dynViewModel.HomeSpaceViewModel as HomeWorkspaceViewModel).Model.RequestRun();
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs:145]: If we hide this dialog correctly, we should not run into case where `Yes` button is clicked on non-HomeSpaceViewModel. That was why I was asking you to hide it since it should not display in the first place. I dont think we need this check here..

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 5]
```diff
@@ -111,7 +111,13 @@ public Point TransformOrigin
 
         public bool ViewingHomespace
         {
-            get { return model.CurrentWorkspace == HomeSpace; }
+            get {
+                if (model.CurrentWorkspace != HomeSpace)
+                {
+                    onNotViewingHomeSpace();
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:117]: Making calls or throwing events in getter usually is not desired. We sometimes do it in setter but since in Dynamo we already have OnPropertyChanged() function ready for this, I would move this to there, there is a switch block to handle property change and what would be the consequences. We just need to flip the visibility based on the value of this property I think.

File: src/DynamoCoreWpf/Controls/NotificationsControl.xaml.cs [hunk 6]
```diff
@@ -40,9 +40,9 @@ void NotificationsControl_Loaded(object sender, RoutedEventArgs e)
         /// <param name="e"></param>
         void window_PreviewMouseDown(object sender, MouseButtonEventArgs e)
         {
-            var hsvm = (HomeWorkspaceViewModel)((DynamoViewModel)DataContext).HomeSpaceViewModel;
+            var hsvm = (HomeWorkspaceViewModel)((DynamoViewModel)DataContext).HomeSpaceViewModel;            
             // Commented this after MAGN - 8423
-            // hsvm.ClearWarning();
+            //hsvm.ClearWarning();
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Controls/NotificationsControl.xaml.cs:45]: Would you revert the change to this file? In general, there should be a space after the `//` for comments

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs (hunk 3 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:278]: I dont think we still need this fix here in addition to other changes?

File: src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs [hunk 7]
```diff
@@ -47,6 +47,11 @@ public FileTrustWarning(DynamoView dynamoViewWindow)
             SetUpPopup();
 
             HomeWorkspaceModel.WorkspaceClosed += CloseWarningPopup;
+            dynViewModel.NotViewingHomeSpace += DynViewModel_NotViewingHomeSpace; ;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs:50]: unsubscribe.. But I think this solution is not general enough since it only handles the `NotViewingHomeSpace` case.. I have mentioned reasoning above.

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs [hunk 8]
```diff
@@ -128,10 +128,18 @@ public HomeWorkspaceViewModel(HomeWorkspaceModel model, DynamoViewModel dynamoVi
             hwm.SetNodeDeltaState +=hwm_SetNodeDeltaState;
 
             dynamoViewModel.Model.ShutdownStarted += Model_ShutdownStarted;
-
+            dynamoViewModel.PropertyChanged += DynamoViewModel_PropertyChanged;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:131]: Thanks, we will also need to unsubscribe it somewhere

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs [hunk 9]
```diff
@@ -128,10 +128,18 @@ public HomeWorkspaceViewModel(HomeWorkspaceModel model, DynamoViewModel dynamoVi
...
             SetupFooterNotificationItems();
         }
 
+        private void DynamoViewModel_PropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == "CurrentSpace" && !((sender as DynamoViewModel).HomeSpaceViewModel as HomeWorkspaceViewModel).IsCurrentSpace)
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:137]: hmm why not just watch `ViewingHomespace`? This seems to be the same check

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs [hunk 10]
```diff
@@ -128,10 +128,18 @@ public HomeWorkspaceViewModel(HomeWorkspaceModel model, DynamoViewModel dynamoVi
...
+            {
+                ClearWarning();
+            }
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:140]: Do we need an else case to make it visible when switching back from custom node workspace?

File: src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs [hunk 11]
```diff
@@ -47,6 +47,15 @@ public FileTrustWarning(DynamoView dynamoViewWindow)
...
             HomeWorkspaceModel.WorkspaceClosed += CloseWarningPopup;
+            dynViewModel.PropertyChanged += DynViewModel_PropertyChanged;
+        }
+
+        private void DynViewModel_PropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == "CurrentSpace" && !((sender as DynamoViewModel).HomeSpaceViewModel as HomeWorkspaceViewModel).IsCurrentSpace)
+            {
+                IsOpen = false;
+            }
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs:58]: Do we need an else case to make it visible again when switching back?

File: src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs [hunk 12]
```diff
@@ -47,6 +47,15 @@ public FileTrustWarning(DynamoView dynamoViewWindow)
             SetUpPopup();
 
             HomeWorkspaceModel.WorkspaceClosed += CloseWarningPopup;
+            dynViewModel.PropertyChanged += DynViewModel_PropertyChanged;
+        }
+
+        private void DynViewModel_PropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == "CurrentSpace" && !((sender as DynamoViewModel).HomeSpaceViewModel as HomeWorkspaceViewModel).IsCurrentSpace)
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs:55]: hmm why not just watch ViewingHomespace? This seems to be the same check

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs [hunk 13]
```diff
@@ -128,10 +128,18 @@ public HomeWorkspaceViewModel(HomeWorkspaceModel model, DynamoViewModel dynamoVi
...
-
+            dynamoViewModel.PropertyChanged += DynamoViewModel_PropertyChanged;
             SetupFooterNotificationItems();
         }
 
+        private void DynamoViewModel_PropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == "CurrentSpace" && !((sender as DynamoViewModel).HomeSpaceViewModel as HomeWorkspaceViewModel).IsCurrentSpace)
+            {
+                ClearWarning();
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:139]: Can you explain to me again why we need to do this? Wouldn't warning be cleared forever after user switching to custom node workspace?

Comment (jesusalvino) [reply src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:139]: That's true for the switching scenario. However the reason why we need that is to hide it immediately the custom node workspace is created (it remains visible from the HomeWorkspace)

File: src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs (hunk 11 above)
Comment (jesusalvino) [reply src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs:58]: No because the HomeworkSpace do that here UpdateRunStatusMsgBasedOnStates , it looks like the trust ui was focused only with one Workspace.

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs (hunk 13 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:139]: @jesusalvino if user switch back from custom node workspace to homeworkspace, would warnings show up again?

**Review by QilongTang — COMMENTED** [review]
Looks much cleaner, LGTM once the remaining comments addressed

File: src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs (hunk 12 above)
Comment (jesusalvino) [reply src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs]: Done

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs (hunk 8 above)
Comment (jesusalvino) [reply src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:131]: Done, also in the FileTrustWarning

File: src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs (hunk 7 above)
Comment (jesusalvino) [reply src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs]: Done

File: src/DynamoCoreWpf/Controls/NotificationsControl.xaml.cs (hunk 6 above)
Comment (jesusalvino) [reply src/DynamoCoreWpf/Controls/NotificationsControl.xaml.cs]: Done

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 5 above)
Comment (jesusalvino) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: Done

File: src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs (hunk 4 above)
Comment (jesusalvino) [reply src/DynamoCoreWpf/Views/FileTrust/FileTrustWarning.xaml.cs]: Done

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs (hunk 13 above)
Comment (jesusalvino) [reply src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:139]: Warnings show up again in the HomeworkSpace only if is required, it's already handled by the HomeworkSpace in fact. For the reported scenario, yes.

**Comment by jesusalvino** [comment]
> Looks much cleaner, LGTM once the remaining comments addressed

Done

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs [hunk 14]
```diff
@@ -128,10 +128,18 @@ public HomeWorkspaceViewModel(HomeWorkspaceModel model, DynamoViewModel dynamoVi
...
 
             dynamoViewModel.Model.ShutdownStarted += Model_ShutdownStarted;
-
+            dynamoViewModel.PropertyChanged += DynamoViewModel_PropertyChanged;
             SetupFooterNotificationItems();
         }
 
+        private void DynamoViewModel_PropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == "CurrentSpace" && !(sender as DynamoViewModel).ViewingHomespace)
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:137]: use `nameof` to avoid spelling mistakes and make refactoring easier.

Comment (jesusalvino) [reply src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs]: > use `nameof` to avoid spelling mistakes and make refactoring easier.

Done

Comment (jesusalvino) [reply src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs]: @QilongTang Kindly, Could you let me know if we have any pending here please ?

**Review by QilongTang — APPROVED** [review]
LGTM

File: src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs [hunk 15]
```diff
@@ -128,10 +128,18 @@ public HomeWorkspaceViewModel(HomeWorkspaceModel model, DynamoViewModel dynamoVi
...
 
             dynamoViewModel.Model.ShutdownStarted += Model_ShutdownStarted;
-
+            dynamoViewModel.PropertyChanged += DynamoViewModel_PropertyChanged;
             SetupFooterNotificationItems();
         }
 
+        private void DynamoViewModel_PropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            if (e.PropertyName == nameof(DynamoViewModel.CurrentSpace) && !(sender as DynamoViewModel).ViewingHomespace)
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs:137]: I think you can check `PropertyName == nameof(DynamoViewModel.ViewingHomespace)` directly but this also works

**Comment by QilongTang** [comment]
Waiting for PR checks and also checking @jesusalvino is the last gif still valid?

**Comment by jesusalvino** [comment]
> Waiting for PR checks and also checking @jesusalvino is the last gif still valid?

yes, the last gif is valid since the behavior is the same with the refactoring

**Comment by QilongTang** [comment]
passed at https://master-15.jenkins.autodesk.com/view/DYN/job/DYN-DevCI_Self_Service/1043
